### PR TITLE
auth: use a retryable http client to handle 429 responses from token endpoint(s)

### PR DESCRIPTION
**Pull request by manicminer**
Docs: https://learn.microsoft.com/en-us/azure/active-directory/fundamentals/resilience-client-app
Related: https://github.com/hashicorp/terraform-provider-azurerm/issues/20901

### CODE REVIEW

File: sdk/auth/client.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import (
+	"net/http"
+
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+// httpClient returns a shimmed retryablehttp Client, at the moment with default retry,
+// backoff and timeout settings, but we may need to change this in the future.
+func httpClient() *http.Client {
+	retryableClient := retryablehttp.NewClient()
```
Comment (tombuildsstuff) [inline sdk/auth/client.go:12]: double-checking this includes the ProxyFromEnvironment setting?

File: sdk/auth/managed_identity_authorizer.go [hunk 2]
```diff
@@ -169,10 +169,10 @@ func azureMetadata(ctx context.Context, url string) (body []byte, err error) {
 	req.Header = http.Header{
 		"Metadata": []string{"true"},
 	}
-	client := &http.Client{
-		Transport: http.DefaultTransport,
-		Timeout:   msiDefaultTimeout,
-	}
+
+	client := httpClient()
+	client.Timeout = msiDefaultTimeout
```
Comment (tombuildsstuff) [inline sdk/auth/managed_identity_authorizer.go:174]: double-checking: are there any other configuration changes we'd need to make to the retryable client given the Managed Identity API is must more limited with regards to retries than resource manager? https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#retry-guidance

**Review by tombuildsstuff — APPROVED** [review]
A couple of questions inline to sanity-check but otherwise 👍

File: sdk/auth/client.go (hunk 1 above)
Comment (manicminer) [reply sdk/auth/client.go:12]: Fortunately it does, thanks to very [usable defaults](https://github.com/hashicorp/go-cleanhttp/blob/02f12f05b908335f65f124c1a7f1ec45f5c42a35/cleanhttp.go#L26-L42) in go-cleanhttp.

File: sdk/auth/managed_identity_authorizer.go (hunk 2 above)
Comment (manicminer) [reply sdk/auth/managed_identity_authorizer.go:174]: Thanks for the docs link, we should probably add a custom Backoff func for this.

**Review by tombuildsstuff — APPROVED** [review]
One thought but otherwise 👍

**Pull request merged**